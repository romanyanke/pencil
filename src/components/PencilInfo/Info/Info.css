.Info-frame {
  height: 0;
  padding-bottom: 100%;
  position: relative;
}

.Info-photo {
  position: absolute;
  display: block;
  width: 100%;
  cursor: zoom-out;
}

.Info-content {
  padding: 1rem;
  background: #f1f1f1;
  color: #242424;
}

.Info-content article {
  line-height: 1.4;
}
