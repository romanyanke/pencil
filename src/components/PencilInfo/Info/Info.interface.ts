import { Pencil } from './../../Pencil/Pencil.interface'

export interface InfoProps {
  pencil: Pencil
}
