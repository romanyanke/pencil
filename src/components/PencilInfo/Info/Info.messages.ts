import { defineMessages } from 'react-intl'

export default defineMessages({
  photo: {
    defaultMessage: '{count, number} фото',
    id: 'Info.photo',
  },
  location: {
    defaultMessage: `{flag} {country}, {city}`,
    id: 'Info.fullLocation',
  },
  unknown: {
    defaultMessage: 'Неизвестно откуда',
    id: 'Info.unknown',
  },
})
