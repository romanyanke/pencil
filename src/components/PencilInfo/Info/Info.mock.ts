import { Pencil } from '../../Pencil/Pencil.interface'

export const mockPencil: Pencil = {
  title: '–†–∏–º—Å–∫–∏–π –ü–∏–Ω–æ–∫–∫–∏–æ',
  geo: 'ITA',
  city: '–†–∏–º',
  id: 'roma',
  grid: 2,
  count: 1,
  country: { flag: 'üáÆüáπ', geo: 'ITA', name: '–ò—Ç–∞–ª–∏—è', pencils: 1 },
  map: {
    lat: 41.892099,
    lng: 12.474878,
  },
  content:
    "<ul> <li><strong>–ü–∏–Ω–æ–∫–∫–∏–æ I</strong></li> <li><a href='?display=pinocchio-second'>–ü–∏–Ω–æ–∫–∫–∏–æ II</a></li> <li><a href='?display=pinocchio-third'>–ü–∏–Ω–æ–∫–∫–∏–æ III</a></li> </ul>  <p>Roma, –†–∏–º. –ù–∞ –∫–æ–Ω—Ü–µ –∫–∞—Ä–∞–Ω–¥–∞—â–∞ —Å–∏–¥–∏—Ç –¥–µ—Ä–µ–≤—è–Ω–Ω—ã–π –ø–∏–Ω–æ–∫–∫–∏–æ, —Ä—É–∫–∏ –∏ –Ω–æ–≥–∏ –±–æ–ª—Ç–∞—é—Ç—Å—è –Ω–∞ –≤–µ—Ä–µ–≤–æ—á–∫–∞—Ö, –∞ –∏–∑ –∫–æ–ª–ø–∞–∫–∞ —Ç–æ—Ä—á–∏—Ç –ø–µ—Ç–ª—è –∏–∑ —Ä–µ–∑–∏–Ω–∫–∏, —á—Ç–æ–±—ã –ø–æ–¥–≤–µ—Å–∏—Ç—å –µ–≥–æ –∫—É–¥–∞-–Ω–∏–±—É–¥—å. –ï—â–µ —É –Ω–µ–≥–æ –µ—Å—Ç—å –¥–≤–∞ –±—Ä–∞—Ç–∞-–±–¥–∏–∑–Ω–µ—Ü–∞.</p> ",
  photos: [
    'https://romanyanke.github.io/pencilbox-static/pencil/roma/gallery/roma-1.jpg',
    'https://romanyanke.github.io/pencilbox-static/pencil/roma/gallery/roma-2.jpg',
    'https://romanyanke.github.io/pencilbox-static/pencil/roma/gallery/roma-3.jpg',
  ],
  preview:
    'https://romanyanke.github.io/pencilbox-static/pencil/roma/gallery/roma-1_hufb33aafe8ffd214e4fc11de6b8dacb8a_28007_300x0_resize_q75_box.jpg',
  tags: [
    '–≥–æ—Å—É–¥–∞—Ä—Å—Ç–≤–µ–Ω–Ω—ã–π',
    '–≥–æ—Ä–æ–¥',
    '—á–µ–ª–æ–≤–µ–∫',
    '—Å –∏–≥—Ä—É—à–∫–æ–π',
    '–∫—Ä—É–≥–ª—ã–π',
    '–±–µ–∑ —Ä–µ–∑–∏–Ω–∫–∏',
    '–Ω–æ–≤—ã–π',
    '—Ç—É–ø–æ–π',
  ],
}
