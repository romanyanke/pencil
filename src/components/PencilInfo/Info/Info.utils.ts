import { isEmpty, pick } from 'lodash'
import qs from 'qs'
import { Pencil } from '../../Pencil/Pencil.interface'
import { Filter } from './../../Filter/Filter.interface'
import { getEmptyFilter } from './../../Filter/Filter.utils'

export const getFilterFromLink = (link: string): Filter | null => {
  const queryPart = link.split('?')[1]
  if (queryPart) {
    const emptyFilter = getEmptyFilter()
    const query = qs.parse(queryPart)
    const filterKeys = Object.keys(emptyFilter) as Array<keyof Filter>
    const keysFromQuery = pick<Filter, keyof Filter>(query, filterKeys)
    if (!isEmpty(keysFromQuery)) {
      return keysFromQuery
    }
  }

  return null
}
