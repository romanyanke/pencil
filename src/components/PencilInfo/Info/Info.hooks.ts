import { useCallback } from 'react'
import { getFilterFromLink } from './Info.utils'
import { useFilter } from '../../Filter/Filter.hooks'
import { getEmptyFilter } from '../../Filter/Filter.utils'

export const usePseudoClick = () => {
  const [, { updateFilter }] = useFilter()
  const handlePseudoLink = useCallback(
    (e: React.MouseEvent<HTMLElement, MouseEvent>) => {
      e.stopPropagation()
      const targetElement = e.target as HTMLElement
      if (targetElement.tagName === 'A') {
        const link = targetElement.getAttribute('href')
        if (link) {
          const filterFromLink = getFilterFromLink(link)
          if (filterFromLink) {
            e.preventDefault()
            updateFilter({ ...getEmptyFilter(), ...filterFromLink })
          }
        }
      }
    },
    [updateFilter],
  )

  return handlePseudoLink
}
