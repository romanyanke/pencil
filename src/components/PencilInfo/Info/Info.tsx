import React from 'react'
import { FormattedMessage } from 'react-intl'
import { usePseudoClick } from './Info.hooks'
import { InfoProps } from './Info.interface'
import messages from './Info.messages'
import { mapFilterToQueryString } from '../../Filter/Filter.utils'

const Info = ({ pencil }: InfoProps) => {
  const handlePseudoLink = usePseudoClick()
  const { country, city } = pencil

  return (
    <div className="Info">
      <div className="Info-content" onClick={handlePseudoLink}>
        <h1>{pencil.title}</h1>

        <h2>
          {city && country ? (
            <FormattedMessage
              {...messages.location}
              values={{
                flag: country.flag,
                country: country.name,
                city: pencil.city,
              }}
            />
          ) : city ? (
            city
          ) : (
            <FormattedMessage {...messages.unknown} />
          )}
        </h2>

        <article dangerouslySetInnerHTML={{ __html: pencil.content }} />

        <p>
          <FormattedMessage
            tagName="b"
            {...messages.photo}
            values={{ count: pencil.photos.length }}
          />
        </p>
      </div>

      {pencil.photos.map(src => (
        <div key={src} className="Info-frame">
          <img className="Info-photo" alt={pencil.title} src={src} />
        </div>
      ))}

      <div className="Info-content" onClick={handlePseudoLink}>
        <article className="Info-tags">
          🏷
          {pencil.tags.map(tag => (
            <a className="Info-tag" key={tag} href={mapFilterToQueryString({ tag })}>
              {tag}
            </a>
          ))}
        </article>
      </div>
    </div>
  )
}

export default Info
