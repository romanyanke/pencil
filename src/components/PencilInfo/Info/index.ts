import './Info.css'

export { default } from './Info'
