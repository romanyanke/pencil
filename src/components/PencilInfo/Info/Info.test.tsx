import { render, fireEvent } from '@testing-library/react'
import React from 'react'
import { IntlProvider } from 'react-intl'
import { Provider } from 'react-redux'
import { mockPencil } from './Info.mock'
import Info from '.'
import store from '../../../store'

describe('<Info />', () => {
  const renderWithProviders = () => ({
    ...render(
      <Provider store={store}>
        <IntlProvider locale="en" defaultLocale="en">
          <Info pencil={mockPencil} />
        </IntlProvider>
      </Provider>,
    ),
    rerender: renderWithProviders,
    store,
  })

  it('should match snapshot', () => {
    const { container } = renderWithProviders()
    expect(container).toMatchSnapshot()
  })

  it('should update filter when link is cliked', () => {
    const { getByText, store } = renderWithProviders()
    const link = getByText('Пиноккио II')

    expect(store.getState()).toMatchObject({ filter: { display: '' } })
    fireEvent.click(link)
    expect(store.getState()).toMatchObject({ filter: { display: 'pinocchio-second' } })
  })
})
