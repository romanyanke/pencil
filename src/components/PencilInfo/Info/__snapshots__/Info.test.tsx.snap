// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Info /> should match snapshot 1`] = `
<div>
  <div
    class="Info"
  >
    <div
      class="Info-content"
    >
      <h1>
        –†–∏–º—Å–∫–∏–π –ü–∏–Ω–æ–∫–∫–∏–æ
      </h1>
      <h2>
        üáÆüáπ –ò—Ç–∞–ª–∏—è, –†–∏–º
      </h2>
      <article>
        <ul>
           
          <li>
            <strong>
              –ü–∏–Ω–æ–∫–∫–∏–æ I
            </strong>
          </li>
           
          <li>
            <a
              href="?display=pinocchio-second"
            >
              –ü–∏–Ω–æ–∫–∫–∏–æ II
            </a>
          </li>
           
          <li>
            <a
              href="?display=pinocchio-third"
            >
              –ü–∏–Ω–æ–∫–∫–∏–æ III
            </a>
          </li>
           
        </ul>
          
        <p>
          Roma, –†–∏–º. –ù–∞ –∫–æ–Ω—Ü–µ –∫–∞—Ä–∞–Ω–¥–∞—â–∞ —Å–∏–¥–∏—Ç –¥–µ—Ä–µ–≤—è–Ω–Ω—ã–π –ø–∏–Ω–æ–∫–∫–∏–æ, —Ä—É–∫–∏ –∏ –Ω–æ–≥–∏ –±–æ–ª—Ç–∞—é—Ç—Å—è –Ω–∞ –≤–µ—Ä–µ–≤–æ—á–∫–∞—Ö, –∞ –∏–∑ –∫–æ–ª–ø–∞–∫–∞ —Ç–æ—Ä—á–∏—Ç –ø–µ—Ç–ª—è –∏–∑ —Ä–µ–∑–∏–Ω–∫–∏, —á—Ç–æ–±—ã –ø–æ–¥–≤–µ—Å–∏—Ç—å –µ–≥–æ –∫—É–¥–∞-–Ω–∏–±—É–¥—å. –ï—â–µ —É –Ω–µ–≥–æ –µ—Å—Ç—å –¥–≤–∞ –±—Ä–∞—Ç–∞-–±–¥–∏–∑–Ω–µ—Ü–∞.
        </p>
         
      </article>
      <p>
        <b>
          3 —Ñ–æ—Ç–æ
        </b>
      </p>
    </div>
    <div
      class="Info-frame"
    >
      <img
        alt="–†–∏–º—Å–∫–∏–π –ü–∏–Ω–æ–∫–∫–∏–æ"
        class="Info-photo"
        src="https://romanyanke.github.io/pencilbox-static/pencil/roma/gallery/roma-1.jpg"
      />
    </div>
    <div
      class="Info-frame"
    >
      <img
        alt="–†–∏–º—Å–∫–∏–π –ü–∏–Ω–æ–∫–∫–∏–æ"
        class="Info-photo"
        src="https://romanyanke.github.io/pencilbox-static/pencil/roma/gallery/roma-2.jpg"
      />
    </div>
    <div
      class="Info-frame"
    >
      <img
        alt="–†–∏–º—Å–∫–∏–π –ü–∏–Ω–æ–∫–∫–∏–æ"
        class="Info-photo"
        src="https://romanyanke.github.io/pencilbox-static/pencil/roma/gallery/roma-3.jpg"
      />
    </div>
    <div
      class="Info-content"
    >
      <article
        class="Info-tags"
      >
        üè∑
        <a
          class="Info-tag"
          href="?tag=%D0%B3%D0%BE%D1%81%D1%83%D0%B4%D0%B0%D1%80%D1%81%D1%82%D0%B2%D0%B5%D0%BD%D0%BD%D1%8B%D0%B9"
        >
          –≥–æ—Å—É–¥–∞—Ä—Å—Ç–≤–µ–Ω–Ω—ã–π
        </a>
        <a
          class="Info-tag"
          href="?tag=%D0%B3%D0%BE%D1%80%D0%BE%D0%B4"
        >
          –≥–æ—Ä–æ–¥
        </a>
        <a
          class="Info-tag"
          href="?tag=%D1%87%D0%B5%D0%BB%D0%BE%D0%B2%D0%B5%D0%BA"
        >
          —á–µ–ª–æ–≤–µ–∫
        </a>
        <a
          class="Info-tag"
          href="?tag=%D1%81%20%D0%B8%D0%B3%D1%80%D1%83%D1%88%D0%BA%D0%BE%D0%B9"
        >
          —Å –∏–≥—Ä—É—à–∫–æ–π
        </a>
        <a
          class="Info-tag"
          href="?tag=%D0%BA%D1%80%D1%83%D0%B3%D0%BB%D1%8B%D0%B9"
        >
          –∫—Ä—É–≥–ª—ã–π
        </a>
        <a
          class="Info-tag"
          href="?tag=%D0%B1%D0%B5%D0%B7%20%D1%80%D0%B5%D0%B7%D0%B8%D0%BD%D0%BA%D0%B8"
        >
          –±–µ–∑ —Ä–µ–∑–∏–Ω–∫–∏
        </a>
        <a
          class="Info-tag"
          href="?tag=%D0%BD%D0%BE%D0%B2%D1%8B%D0%B9"
        >
          –Ω–æ–≤—ã–π
        </a>
        <a
          class="Info-tag"
          href="?tag=%D1%82%D1%83%D0%BF%D0%BE%D0%B9"
        >
          —Ç—É–ø–æ–π
        </a>
      </article>
    </div>
  </div>
</div>
`;
